src/JsonSchema/Constraints/TypeConstraint.php

```diff
@@ -249,14 +267,46 @@ protected function toNumber($value)
         if (is_numeric($value)) {
             return $value + 0; // cast to number
         }
+        if (is_bool($value) || is_null($value)) {
+            return (int) $value;
+        }
+        if ($this->getTypeCheck()->isArray($value) && count($value) === 1) {
+            reset($value);
+
+            return $this->toNumber(current($value));
+        }
 
         return $value;
     }
 
+    /**
+     * Converts a value to an integer
+     *
+     * @param mixed $value
+     *
+     * @return mixed
+     */
     protected function toInteger($value)
     {
-        if (is_numeric($value) && (int) $value == $value) {
-            return (int) $value; // cast to number
+        $numberValue = $this->toNumber($value);
+        if (is_numeric($numberValue) && (int) $numberValue == $numberValue) {
+            return (int) $numberValue; // cast to number
+        }
+
+        return $value;
+    }
+
+    /**
+     * Converts a value to an array containing that value
+     *
+     * @param mixed $value
+     *
+     * @return array
```

Comment:
I like your original suggestion better. If we're not using PHPdoc, then there's no reason to avoid it IMO. If such a reason crops up later, it's a pretty quick / easy change to downgrade the return type to `mixed`.